packages/neuron-wallet/src/services/tx/transaction-generator.ts

```diff
@@ -1332,6 +1335,131 @@ export class TransactionGenerator {
     return tx
   }
 
+  public static async generateSudtMigrateAcpTx(
+    sudtCell: Output,
+    acpAddress?: string,
+    fee: string = '0',
+    feeRate: string = '1000'
+  ) {
+    const currentWallet = WalletService.getInstance().getCurrent()
+    if (!currentWallet) {
+      throw new CurrentWalletNotSet()
+    }
+    const assetAccountInfo = new AssetAccountInfo()
+    const sudtMigrateAcpInput = [
+      Input.fromObject({
+        previousOutput: sudtCell.outPoint!,
+        capacity: sudtCell.capacity,
+        lock: sudtCell.lock,
+        type: sudtCell.type,
+        lockHash: sudtCell.lockHash,
+        data: sudtCell.data,
+        since: '0'
+      })
+    ]
+
+    const secpCellDep = await SystemScriptInfo.getInstance().getSecpCellDep()
+    const sudtCellDep = assetAccountInfo.sudtCellDep
+    const anyoneCanPayDep = assetAccountInfo.anyoneCanPayCellDep
+    let outputs: Output[] = []
+    if (acpAddress) {
+      if (!sudtCell.type) {
+        throw new MigrateSudtCellNoTypeError()
+      }
+      const receiverAcpCells = await CellsService.getACPCells(
+        Script.fromSDK(addressToScript(acpAddress)),
+        sudtCell.type
+      )
+      if (!receiverAcpCells.length) {
+        throw new TargetOutputNotFoundError()
+      }
+      const receiverAcpCell = receiverAcpCells[0]
+      const receiverAcpInputAmount = BufferUtils.readBigUInt128LE(receiverAcpCell.data)
+      const sudtCellAmount = BufferUtils.readBigUInt128LE(sudtCell.data)
+      const receiverAcpOutputAmount = receiverAcpInputAmount + sudtCellAmount
+      sudtCell.setData('0x')
+      sudtCell.setType(null)
+      outputs = [
+        sudtCell,
+        Output.fromObject({
+          capacity: receiverAcpCell.capacity,
+          lock: receiverAcpCell.lockScript(),
+          type: receiverAcpCell.typeScript(),
+          data: BufferUtils.writeBigUInt128LE(receiverAcpOutputAmount)
+        })
+      ]
+      const receiverAcpCellModel = receiverAcpCell.toModel()
+      sudtMigrateAcpInput.push(
+        Input.fromObject({
+          previousOutput: receiverAcpCellModel.outPoint!,
+          capacity: receiverAcpCellModel.capacity,
+          lock: receiverAcpCellModel.lock,
+          type: receiverAcpCellModel.type,
+          lockHash: receiverAcpCellModel.lockHash,
+          data: receiverAcpCellModel.data,
+          since: '0'
+        })
+      )
+    } else {
+      const addresses = await currentWallet.getNextReceivingAddresses()
+      const usedBlake160s = new Set(await AssetAccountService.blake160sOfAssetAccounts())
```

Comment:
This can be called only on `currentWallet.isHDWallet() === true`

But I don't understand why HDWallet and non-HDWallet have different logic

Reply:
Because it's same as create new sudt account. I refer to [create new account code](https://github.com/nervosnetwork/neuron/blob/35e8cf8949d33971d91b1bb98fe563e45ef21149/packages/neuron-wallet/src/services/asset-account-service.ts#L197-L199)

Reply:
Gotcha

Reply:
`usedBlake160s` is not used when `currentWallet.isHDWallet()` is truthy so reading from database could be avoided.